lib/broadway.ex

```diff
@@ -979,6 +979,13 @@ defmodule Broadway do
         raise ArgumentError, format_error(error)
 
       {:ok, opts} ->
+        Enum.each(opts[:batchers], fn {batcher_name, batcher_opt} ->
+          if batcher_opt[:max_demand] == nil and is_tuple(batcher_opt[:batch_size]) do
+            raise ArgumentError,
+                  "For batch #{batcher_name}, expected option :max_demand to be provided when :batch_size is a tuple"
+          end
+        end)
+
```

Comment:
Would it be better to raise the argument error earlier instead of wait part of Broadway process started?  That is why I move the checking from BatcherStage here.

Reply:
The earlier the best as it will lead to clearer error messages.